core-kotlin-modules/core-kotlin-concurrency-3/src/test/java/com/baeldung/singleRxJavaToCoroutineDeferred/SingleRxJavaToCoroutineDeferredUnitTest.kt

```diff
@@ -0,0 +1,216 @@
+            onError(completableDeferred, error)
+        })
+        return completableDeferred
+    }
+
+    @Test
+    fun `test using CompletableDeferred extension`() = runBlocking {
+        val deferred = getFilteredProducts().toCompletableDeferred()
+        deferred.assertResultsTrue()
+    }
+
+    @Test
+    fun `test using CompletableDeferred with callback`(): Unit = runBlocking {
+        getFilteredProducts().toCompletableDeferred(onSuccess = { deferredResult, _ ->
+            launch { deferredResult.assertResultsTrue() }
+        }, onError = { _, error ->
+            println("Error: ${error.message}")
+        }).await()
+    }
+
+    // using suspendCoroutines directly
+    @Test
+    fun `using suspendCoroutines directly`(): Unit = runBlocking {
+        val deffered = async {
+            suspendCoroutine { continuation ->
+                getFilteredProducts().subscribe { result ->
+                    continuation.resume(result)
+                }
+            }
+        }
+        deffered.assertResultsTrue()
+    }
+
+    // using suspendCoroutines with extension
+    private suspend fun <T : Any> Single<T>.toDeferredWithSuspend(): Deferred<T> {
+        return coroutineScope {
+            async {
+                suspendCoroutine { continuation ->
+                    this@toDeferredWithSuspend.subscribe { result ->
+                        continuation.resume(result)
+                    }
+                }
+            }
+        }
+    }
+
+    @Test
+    fun `test using suspendCoroutine`() = runBlocking {
+        val deferred = getFilteredProducts().toDeferredWithSuspend()
+        deferred.assertResultsTrue()
+    }
+
+    // using suspendCancellableCoroutine
+    private suspend fun <T : Any> Single<T>.toDeferredWithsuspendCancellableCoroutine(
+        onSuccess: (Deferred<T>) -> Unit, onError: (Throwable) -> Unit
+    ): Deferred<T> {
+        return coroutineScope {
+            async {
+                suspendCancellableCoroutine { continuation ->
+                    this@toDeferredWithsuspendCancellableCoroutine.subscribe({ result ->
+                        val deferredResult = CompletableDeferred<T>().apply {
+                            complete(result)
+                            continuation.resume(result)
+                        }
+                        onSuccess(deferredResult)
+                    }, { error ->
+                        continuation.resumeWithException(error)
+                        onError(error)
+                    })
+                }
+            }
+        }
+    }
+
+
+    @Test
+    fun `test using suspendCancellableCoroutine with custom callback`(): Unit = runBlocking {
+        getFilteredProducts().toDeferredWithsuspendCancellableCoroutine(onSuccess = { deferredResult ->
+            launch { deferredResult.assertResultsTrue() }
+        }, onError = { error ->
+            println("Error: ${error.message}")
```

Comment:
Don't forget the guideline that we can't use println and should use SLF4J for logging